### PR TITLE
[Caffe2] Some small changes to InferBlobShapesAndTypes definition and SameAsInput Schema

### PR DESCRIPTION
**Pull request by Swetko**

### CODE REVIEW

**Comment by Swetko** [comment]
This is a repeat of this PR: [https://github.com/pytorch/pytorch/pull/6277](url).
Follow discussions there for more info.

File: caffe2/python/pybind_state.cc [hunk 1]
```diff
@@ -1219,13 +1219,15 @@ void addGlobalMethods(py::module& m) {
 
         // Parse protobuffers to NetDefs
         std::vector<std::unique_ptr<caffe2::NetDef>> nets;
```
[the remark this answers is not in the document]
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: Yes, That could work. Will update it in a second. Thank you!

[the remark this answers is not in the document]
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: I tried implementing it like that, but it seems it makes the code longer and less clear necessitating two for loops because you can't take pointers from a vector unless you know its size is not going to change (hence invalidate the pointers).
Here's how it can look. I think it looks worse and is more prone to errors.

        std::vector<caffe2::NetDef> nets;
        std::vector<caffe2::NetDef*> nets_ptr;
        for (auto proto : net_protos) {
          NetDef def;
          CAFFE_ENFORCE(def.ParseFromString(proto));
          nets.push_back(def);
        }
        for (auto& net: nets) nets_ptr.push_back(&net);

[the remark this answers is not in the document]
Comment (yinghai) [reply caffe2/python/pybind_state.cc]: Try something like this? 
```
std::vector<caffe2::NetDef> nets;
    std::vector<caffe2::NetDef*> nets_ptr;
    for (const auto& proto : net_protos) {
      nets.emplace_back();
      auto& net = nets.back();
      CAFFE_ENFORCE(net.ParseFromString(proto));
      nets_ptr.push_back(&net);
    }
```

[the remark this answers is not in the document]
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: This will segfault. When you push/emplace_back elements to the std::vector, it dynamically reallocates its memory when its current capacity fills, meaning any pointers to its old content are invalidated. A way to solve this problem is to preallocate:

std::vector<caffe2::NetDef> nets(net_protos.size());

Bu then we have to iterate with a counter variable like:

for( size_t i=0; i < net_protos.size(); i++)
  {
  nets[i].ParseXXX()
  nets_ptr.push_back(&net[i])
  }

Or alternatively reserve the capacity 

std::vector<caffe2::NetDef> nets;
nets.reserve(net_protos.size());

But this is hacky and bug prone to anyone looking at the function later. Relying on that behavior seems not the most elegant solution to me.

[the remark this answers is not in the document]
Comment (yinghai) [reply caffe2/python/pybind_state.cc]: Yes, reserve will work.

[the remark this answers is not in the document]
Comment (yinghai) [reply caffe2/python/pybind_state.cc]: On the other hand, it shows how dangerous a raw pointer is. :)

[the remark this answers is not in the document]
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: I didn't like the idea of passing pointers in the first place, but I wanted to make minimal changes to the existing code. I'd just change the interface to InferTensorShapes to take reference to a vector of NetDefs. This way no pointers at all. Does that sound better? Right now these functions are only used in this context so it'll be small change. To be honest I don't care too much about this change but when I was exploring the functionalities of Caffe2 I wanted to call InferTensorShapes and found out it's impossible to call the function , unless you explicitly create a NetDef object  (by copying/cloning, etc.) just for the purpose of calling a function that inspects the data. I found that to be bad style so I tried the make the minimal changes.

By the way, what's the problem with the unique_ptr's that is currently implemented in my PR?  The reason for the segfault is because I had the std::move(ptr) before ptr.get()  which was invalidating the pointer. The tests pass now without the segfault.

Let me know which solution you'd prefer and I'll do that.

[the remark this answers is not in the document]
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: shared_ptr would have the same problem as unique_ptr. Those smart pointers are concerned with ownership. In the current context we are talking about the interface of the InferTensorShape function and that function doesn't have anything to do with the ownership of the object. It just does operation on them.
[https://stackoverflow.com/questions/6675651/when-should-i-use-raw-pointers-over-smart-pointers?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa](url)

The proper way to have the interface would be  const T& or const T*, with reference preferable, but in this case it's complicated because we are passing a vector of those. So options are:

vector<T>&  (this would require creating the vector and copying of data, in case your objects are not already stored in a vector. Since this is not a performance critical place, it's a fine solution).

vector<T*> (what I changed it to be, requires creation of just a vector of lightweight pointers)

vector<unique_ptr<T>> ( how it was before, has the same disadvantage of having to copy/create objects just so we pass them to the function, but the idea is less clear and more verbose)

vector<T&> (C++ doesn't allow vector of references).

None of those options is inherently risky. The only issue is how clear is the code, and the more clear the less space for bugs. The option of passing vector<shared_ptr<T>> is actually the least clear. Let me explain:

Suppose we have objects (a and b):

T a,b;

We want to pass them to this function F:

void F(vector<shared_ptr<T>>);

we can call like this:

F({make_shared(&a),make_shared(&b)})

Upon exit it will segfault because destroying the temporary vector<shared_ptr> would cause destruction of stack allocated objects.

Or we can call like this:

F({make_shared(new T(a)),make_shared(new T(b))})

in this case we are creating new objects just for the purpose of calling the function.
Alternatively there's also weak_ptr but that is used in conjunction with shared_ptr.
Basically we are going to have this problem with any function that expects a smart pointer as an input as outlined in the stack overflow link.


The Caffe2 codebase is filled with all possible combinations of references, pointers, and smart pointers. Not sure if there's official guidelines about coding styles in caffe2.

So let me know which interface you'd prefer to use and I'll make the calling functions respect that.

**Review by yinghai — APPROVED** [review]
Ok. Let's go with raw pointers for now.

**Comment by yinghai** [comment]
Thanks for contributing!

**Comment by Swetko** [comment]
Thank you! I would contribute more. I am wondering, if I am thinking of making a change or adding a feature, is there a place where I can say the idea and discuss before making a PR? Is the issues sections for that? Basically I am adding whatever features I need as I go about my work.

**Comment by yinghai** [comment]
Yes, you can just raise an issue in github and you can cross-reference issues and prs by putting their urls in the comments.
